### DISCUSSION TITLE
Exporting Milvus 2.0 data and importing it to another Milvus 2.0

### DISCUSSION THREAD

**Discussion by pwichmann** [milvus-io/milvus#15860] [Q&A and General discussion]
I have a Milvus 2.0 standalone running on one machine and need to get that data transferred into another Milvus 2.0 running on another machine.
How do I best transfer the data (and keep the IDs of all vectors identical)?
The MilvusDM appears not to work for Milvus 2.0.

**Comment by pwichmann** [comment]
Anyone?

**Reply by xiaofan-luan** [reply]
We have a plan to work on that but currently it seems to be not ready yet.
Are you using auto increment primary key?

**Reply by pwichmann** [reply]
Thank you @xiaofan-luan!

I am using an auto-increment primary key. But I need to make sure that the keys remain the same when I transfer the data into a different Milvus instance on another machine. The reason for that is that I reference these Milvus keys in another (non-vector) database.

One obvious option I see is that I use a custom python script to download all tens of millions of vectors from one instance, dump them into a file, and then use another script to upload them again to another instance. But this would be slow and I don't know if I can ensure that primary keys stay the same. Otherwise, I have to update all references to them as well.

Or is it possible to just zip all the milvus/volumes content and copy it into the same folder of the other (currently empty) milvus instance?

**Reply by pwichmann** [reply]
Needless to say: Exporting / importing data is super crucial for databases. You probably know that better than me.

**Reply by xiaofan-luan** [reply]
> Needless to say: Exporting / importing data is super crucial for databases. You probably know that better than me.

Yep you are right, we have a milestone to support data import in 2.1, and export in 2.2(July 2022)

**Reply by xiaofan-luan** [reply]
> Thank you @xiaofan-luan!
> 
> I am using an auto-increment primary key. But I need to make sure that the keys remain the same when I transfer the data into a different Milvus instance on another machine. The reason for that is that I reference these Milvus keys in another (non-vector) database.
> 
> One obvious option I see is that I use a custom python script to download all tens of millions of vectors from one instance, dump them into a file, and then use another script to upload them again to another instance. But this would be slow and I don't know if I can ensure that primary keys stay the same. Otherwise, I have to update all references to them as well.
> 
> Or is it possible to just zip all the milvus/volumes content and copy it into the same folder of the other (currently empty) milvus instance?

If you are using standalone, copy the folder of etcd, rocksdb folder and minio folder should work.  

If would be even easier if you don't want to move etcd and minio, then you can just stop milvus, copy rocksdb and restart milvus on new server, pointing out your etcd and minio back

**Reply by pwichmann** [reply]
Thank you. This could at least be a temporary solution and save quite a bit of time.

**Reply by pwichmann** [reply]
Also, I am very thankful for Milvus. It enables some amazing services I can build on top. Huge fan. <3

**Reply by xiaofan-luan** [reply]
> ge fan. <3

Milvus is still in it's early stage~ Join us if you are interested!
The team is trying our best to deliver openbox vector search and we will support data import, export soon as possible

**Reply by PeterPilley** [reply]
> > Thank you @xiaofan-luan!
> > I am using an auto-increment primary key. But I need to make sure that the keys remain the same when I transfer the data into a different Milvus instance on another machine. The reason for that is that I reference these Milvus keys in another (non-vector) database.
> > One obvious option I see is that I use a custom python script to download all tens of millions of vectors from one instance, dump them into a file, and then use another script to upload them again to another instance. But this would be slow and I don't know if I can ensure that primary keys stay the same. Otherwise, I have to update all references to them as well.
> > Or is it possible to just zip all the milvus/volumes content and copy it into the same folder of the other (currently empty) milvus instance?
> 
> If you are using standalone, copy the folder of etcd, rocksdb folder and minio folder should work.
> 
> If would be even easier if you don't want to move etcd and minio, then you can just stop milvus, copy rocksdb and restart milvus on new server, pointing out your etcd and minio back

Just saw this message, this answers my question in relation to a simple workaround thanks @xiaofan-luan

**Reply by mattbui** [reply]
> > Thank you @xiaofan-luan!
> > I am using an auto-increment primary key. But I need to make sure that the keys remain the same when I transfer the data into a different Milvus instance on another machine. The reason for that is that I reference these Milvus keys in another (non-vector) database.
> > One obvious option I see is that I use a custom python script to download all tens of millions of vectors from one instance, dump them into a file, and then use another script to upload them again to another instance. But this would be slow and I don't know if I can ensure that primary keys stay the same. Otherwise, I have to update all references to them as well.
> > Or is it possible to just zip all the milvus/volumes content and copy it into the same folder of the other (currently empty) milvus instance?
> 
> If you are using standalone, copy the folder of etcd, rocksdb folder and minio folder should work.
> 
> If would be even easier if you don't want to move etcd and minio, then you can just stop milvus, copy rocksdb and restart milvus on new server, pointing out your etcd and minio back

did this work for anyone? I'm using milvus v2.1 and copied everything in milvus/volumes directory to a new server and start milvus pointing to that directory, but collections are listed as empty

**Reply by yhmo** [reply]
> > > Thank you @xiaofan-luan!
> > > I am using an auto-increment primary key. But I need to make sure that the keys remain the same when I transfer the data into a different Milvus instance on another machine. The reason for that is that I reference these Milvus keys in another (non-vector) database.
> > > One obvious option I see is that I use a custom python script to download all tens of millions of vectors from one instance, dump them into a file, and then use another script to upload them again to another instance. But this would be slow and I don't know if I can ensure that primary keys stay the same. Otherwise, I have to update all references to them as well.
> > > Or is it possible to just zip all the milvus/volumes content and copy it into the same folder of the other (currently empty) milvus instance?
> > 
> > 
> > If you are using standalone, copy the folder of etcd, rocksdb folder and minio folder should work.
> > If would be even easier if you don't want to move etcd and minio, then you can just stop milvus, copy rocksdb and restart milvus on new server, pointing out your etcd and minio back
> 
> did this work for anyone? I'm using milvus v2.1 and copied everything in milvus/volumes directory to a new server and start milvus pointing to that directory, but collections are listed as empty

@mattbui Please refer to this thread: https://github.com/milvus-io/milvus/discussions/22300

**Reply by justrp** [reply]
> > Needless to say: Exporting / importing data is super crucial for databases. You probably know that better than me.
> 
> Yep you are right, we have a milestone to support data import in 2.1, and export in 2.2(July 2022)

hello @xiaofan-luan , could you please share if it was implemented? I could not find related documentation. The closest I've seen is https://milvus.io/docs/cli_commands.md which mentions `export` as a supported operation, but seems that it doesn't exist yet.

**Reply by yhmo** [reply]
> > > Needless to say: Exporting / importing data is super crucial for databases. You probably know that better than me.
> > 
> > 
> > Yep you are right, we have a milestone to support data import in 2.1, and export in 2.2(July 2022)
> 
> hello @xiaofan-luan , could you please share if it was implemented? I could not find related documentation. The closest I've seen is https://milvus.io/docs/cli_commands.md which mentions `export` as a supported operation, but seems that it doesn't exist yet.

A backup/restore tool was provided: https://github.com/zilliztech/milvus-backup

**Comment by pwichmann** [comment]
@xiaofan-luan Do the /insert_log/ files in the folder milvus and the folder minio be preserved and also transferred to the other machine?
These appear to consume many, many GB.

**Reply by xiaofan-luan** [reply]
> @xiaofan-luan Do the /insert_log/ files in the folder milvus and the folder minio be preserved and also transferred to the other machine? These appear to consume many, many GB.

insert_log contains all insert datas (after flush)
delta_log contains all deleted datas

**Comment by PeterPilley** [comment]
+1 for this. We are trying to work out what is the best way to achieve all this as well.

Looking at milvusdm documentation it looks like you could achieve it by doing milvus->hdfs and then hdfs->milvus, has anyone tried this?

**Reply by yhmo** [reply]
> +1 for this. We are trying to work out what is the best way to achieve all this as well.
> 
> Looking at milvusdm documentation it looks like you could achieve it by doing milvus->hdfs and then hdfs->milvus, has anyone tried this?

I didn't hear this function before, could you show me the doc which intro milvus->hdfs and hdfs->milvus?
Now we are building a tool for milvus backup: https://github.com/zilliztech/milvus-backup, the work just begin. This tool can export milvus data to minio, and then import the data into another milvus service.

**Reply by PeterPilley** [reply]
Sorry meant to say hdf5 not hdfs.

https://milvus.io/docs/v2.1.x/h2m.md

if all the data is in minio can we just do an export from there and import the data into the new minio manually in the meantime? would that work

**Reply by xiaofan-luan** [reply]
> Sorry meant to say hdf5 not hdfs.
> 
> https://milvus.io/docs/v2.1.x/h2m.md
> 
> if all the data is in minio can we just do an export from there and import the data into the new minio manually in the meantime? would that work

You can do backup and restore , check whether https://github.com/zilliztech/milvus-backup  is what you want

**Reply by Aquar14** [reply]
> +1 for this. We are trying to work out what is the best way to achieve all this as well.
> 
> Looking at milvusdm documentation it looks like you could achieve it by doing milvus->hdfs and then hdfs->milvus, has anyone tried this?

Excuse me.I am trying to export data from Milvus v2.3.1 using MilvusDM and achieve it by doing milvus->hdf5 and then hdf5->milvus. But when i go through milvus->hdf5, the read_milvus_meta.py report a sqlite error: connect failed with unable to open database file. Finally i turn to the official document and find MilvusDM only support data export from milvus 0.x and 1.x.
So i want to know if there is a way to export data from milvus v2.x and save the data as localfile.Thank you.

**Reply by yhmo** [reply]
@Aquar14 
We don't have official tool to export data from milvus v2.x to save as hdf5.
You can use queryIterator to fetch the data batch by batch and save as any format file you want.
https://milvus.io/docs/with-iterators.md#Query-with-an-iterator

**Reply by Aquar14** [reply]
> @Aquar14 We don't have official tool to export data from milvus v2.x to save as hdf5. You can use queryIterator to fetch the data batch by batch and save as any format file you want. https://milvus.io/docs/with-iterators.md#Query-with-an-iterator

Thank you for the reply. I will try it.
